pgjdbc/src/main/java/org/postgresql/core/ParameterContext.java

```diff
@@ -0,0 +1,202 @@
+
+    if (placeholderAtPosition == null) {
+      placeholderAtPosition = new ArrayList<>();
+    }
+
+    placeholderAtPosition.add(bindIndex);
+
+    return bindIndex + 1;
+  }
+
+  public boolean hasParameters() {
+    return placeholderPositions != null && !placeholderPositions.isEmpty();
+  }
+
+  public boolean hasNamedParameters() {
+    return placeholderNames != null && !placeholderNames.isEmpty();
+  }
+
+  /**
+   * @param i 0-indexed position in the order of first appearance
+   * @return The name of the placeholder at this backend parameter position
+   */
+  public String getPlaceholderName(@NonNegative int i) {
+    if (placeholderNames == null) {
+      throw new IllegalStateException("Call hasNamedParameters() first.");
+    }
+    return placeholderNames.get(i);
+  }
+
+  /**
+   * @param i 0-indexed position in the order of appearance
+   * @return The position of the placeholder in the SQL text for this placeholder index
+   */
+  public int getPlaceholderPosition(@NonNegative int i) {
+    if (placeholderPositions == null) {
+      throw new IllegalStateException("Call hasParameters() first.");
+    }
+    return placeholderPositions.get(i);
+  }
+
+  /**
+   * @param i 0-indexed position in the order of appearance
+   * @return The position of the placeholder in the order of first appearance of each placeholder
+   */
+  public int getPlaceholderAtPosition(@NonNegative int i) {
+    if (placeholderAtPosition == null || placeholderAtPosition.isEmpty()) {
+      throw new IllegalStateException("Call hasParameters() first.");
+    }
+    return placeholderAtPosition.get(i);
+  }
+
+  public int getLastPlaceholderPosition() {
+    if (placeholderPositions == null || placeholderPositions.isEmpty()) {
+      throw new IllegalStateException("Call hasParameters() first.");
+    }
+    return placeholderPositions.get(placeholderPositions.size() - 1);
+  }
+
+  /**
+   * Adds a named parameter to this ParameterContext. Once a named Parameter have been added all
+   * subsequent parameters must be named. Using named parameters enable reuse of the same parameters
+   * in several locations of the SQL text. The parameters only have to be sent to the PostgreSQL
+   * backend once per name specified. The values will be sent in the order of the first appearance
+   * of their placeholder.
+   *
+   * @param position in the SQL text where the parser captured the placeholder.
+   * @param bindName is the placeholder name captured by the parser.
+   * @return 1-indexed position in the order of first appearance of named parameters
+   * @throws SQLException if positional and named parameters are mixed.
+   */
+  public int addNamedParameter(@NonNegative int position, String bindName) throws SQLException {
+    checkAndSetBindStyle(BindStyle.NAMED);
+    checkAndAddPosition(position);
+
+    if (placeholderNames == null) {
+      placeholderNames = new ArrayList<>();
+    }
+    int bindIndex = placeholderNames.indexOf(bindName);
```

Comment:
Currently, PostgreSQL supports up to ~16K binds. Could you benchmark the parser at least up to 16K bind names?

I expect this `indexOf` would scan all the contents which looks like `O(N^2)` :(